Connect to Kafka PL/PSC service from TiDB Cloud (#19323)
pingcap/docs: tidb-cloud/changefeed-sink-to-apache-kafka.md

- Currently, TiDB Cloud does not support uploading self-signed TLS certificates to connect to Kafka brokers.
- Because TiDB Cloud uses TiCDC to establish changefeeds, it has the same [restrictions as TiCDC](https://docs.pingcap.com/tidb/stable/ticdc-overview#unsupported-scenarios).
- If the table to be replicated does not have a primary key or a non-null unique index, the absence of a unique constraint during replication could result in duplicated data being inserted downstream in some retry scenarios.

> - If you select **Private Link** or **Private Service Connect** as network connectivity method, please make sure version of TiDB cluster satisfy following conditions.
>   - For 6.5.x, require >= 6.5.9 
>   - For 7.1.x, require >= 7.1.4 
>   - For 7.5.x, require >= 7.5.1 
>   - Support all versions of 8.1.x and later

Reviewer: ```suggestion
- If you select **Private Link** or **Private Service Connect** as network connectivity method, make sure the version of TiDB cluster satisfies the following conditions.

   - For 6.5.x, require v6.5.9 or later
   - For 7.1.x, require v7.1.4 or later
   - For 7.5.x, require v7.5.1 or later
   - Support all versions of v8.1.x and later
```

> - Currently, TiDB Cloud does not support uploading self-signed TLS certificates to connect to Kafka brokers.

Reviewer: Preview https://pingcap-docsite-preview-hvfal0j3g-hfxsds-projects.vercel.app/tidbcloud/changefeed-sink-to-apache-kafka

- Because TiDB Cloud uses TiCDC to establish changefeeds, it has the same [restrictions as TiCDC](https://docs.pingcap.com/tidb/stable/ticdc-overview#unsupported-scenarios).
- If the table to be replicated does not have a primary key or a non-null unique index, the absence of a unique constraint during replication could result in duplicated data being inserted downstream in some retry scenarios.

>     - If you choose Private Link or Private Service Connect as the network connectivity method, ensure that your TiDB cluster version meets the following requirements:
> 
>         - For v6.5.x: version v6.5.9 or later
>         - For v7.1.x: version v7.1.4 or later
>         - For v7.5.x: version v7.5.1 or later
>         - For v8.1.x: all versions of v8.1.0 and later are supported

Reviewer: ```suggestion
- If you choose Private Link or Private Service Connect as the network connectivity method, ensure that your TiDB cluster version meets the following requirements:

    - For v6.5.x: version v6.5.9 or later
    - For v7.1.x: version v7.1.4 or later
    - For v7.5.x: version v7.5.1 or later
    - For v8.1.x: all versions of v8.1.0 and later are supported
```

> - If you select **Private Link** or **Private Service Connect** as network connectivity method, make sure that your TiDB cluster version meets the following conditions:
> 
>     - For v6.5.x, the version must be v6.5.9 or later
>     - For v7.1.x, the version must be v7.1.4 or later
>     - For v7.5.x, the version must be v7.5.1 or later
>     - For v8.1.x, support all versions of v8.1.0 and later

Reviewer: ```suggestion
    - If you choose Private Link or Private Service Connect as the network connectivity method, ensure that your TiDB cluster version meets the following requirements:

        - For v6.5.x: version v6.5.9 or later
        - For v7.1.x: version v7.1.4 or later
        - For v7.5.x: version v7.5.1 or later
        - For v8.1.x: all versions of v8.1.0 and later are supported
```

### Network

Make sure that your TiDB cluster can connect to the Apache Kafka service. There are 3 kinds of network connection methods can be used to connect to Kafka.
1. Private Connect
2. VPC Peering
3. Public IP

If you want a quick try, you can choose **Public IP**. If you want cost-effective, you can choose **VPC Peering**, trade off VPC CIDR conflict and security. If you want to get rid of VPC CIDR conflict and satisfy security compliance, **Private Connect** is the choice, but it will introduce extra [Private Data Link Cost](/tidbcloud/tidb-cloud-billing-ticdc-rcu.md#private-data-link-cost)

#### Private Connect
Private Connect leverages **Private Link** or **Private Service Connect** technologies which provided by cloud vendors, that allow the resources in your VPC to connect to services in other VPCs using private IP addresses, as if those services were hosted directly in your VPC. 

Currently, we only support Private Connect to self-hosted Kafka.

> 1. If your Apache Kafka service already or will be setup in AWS, please follow [Setup Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) to make sure the network connection is set up properly.

Author: output Endpoint Service Name .....

### Network

> Make sure that your TiDB cluster can connect to the Apache Kafka service. There are 3 kinds of network connection methods can be used to connect to Kafka.
> 1. Private Connect
> 2. VPC Peering
> 3. Public IP
> 
> If you want a quick try, you can choose **Public IP**. If you want cost-effective, you can choose **VPC Peering**, trade off VPC CIDR conflict and security. If you want to get rid of VPC CIDR conflict and satisfy security compliance, **Private Connect** is the choice, but it will introduce extra [Private Data Link Cost](/tidbcloud/tidb-cloud-billing-ticdc-rcu.md#private-data-link-cost)

Reviewer: ```suggestion
Ensure that your TiDB cluster can connect to the Apache Kafka service. You can choose one of the following connection methods:

- Private Connect: ideal for avoiding VPC CIDR conflicts and meeting security compliance, but incurs additional [Private Data Link Cost](/tidbcloud/tidb-cloud-billing-ticdc-rcu.md#private-data-link-cost).
- VPC Peering: suitable for a cost-effective solution, but requires managing potential VPC CIDR conflicts and security considerations.
- Public IP: suitable for a quick setup.
```

> #### Private Connect
> Private Connect leverages **Private Link** or **Private Service Connect** technologies which provided by cloud vendors, that allow the resources in your VPC to connect to services in other VPCs using private IP addresses, as if those services were hosted directly in your VPC. 
> 
> Currently, we only support Private Connect to generic Kafka. No special integration with MSK and Confluent Kafka, or any others.
> 1. If your Apache Kafka service already or will be setup in AWS, please follow [Setup Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) to make sure the network connection is set up properly. After set up done, you will get following information back to TiDB Cloud console to create changefeed:
>    - AZs of Kafka Endpoint Service
>    - ID in Kafka Advertised Listener Pattern
>    - Endpoint Service Name
>    - Bootstrap Ports
> 2. If your Apache Kafka service already or will be setup in Google Cloud, please follow [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) to make sure the network connection is set up properly. After set up done, you will get following information back to TiDB Cloud console to create changefeed:
>    - ID in Kafka Advertised Listener Pattern
>    - Service Attachment
>    - Bootstrap Ports

Reviewer: ```suggestion
#### Private Connect

Private Connect leverages **Private Link** or **Private Service Connect** technologies which provided by cloud vendors, that allow the resources in your VPC to connect to services in other VPCs using private IP addresses, as if those services were hosted directly in your VPC. 

Currently, TiDB Cloud only supports Private Connect to generic Kafka. No special integration with MSK and Confluent Kafka, or any others.

- If your Apache Kafka service already or will be set up in AWS, follow [Setup Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) to make sure the network connection is set up properly. After you set it up, you can get the following information back to the TiDB Cloud console to create the changefeed:

    - AZs of Kafka Endpoint Service
    - ID in Kafka Advertised Listener Pattern
    - Endpoint Service Name
    - Bootstrap Ports

- If your Apache Kafka service already or will be set up in Google Cloud, follow [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) to make sure the network connection is set up properly. After you set it up, you can get the following information back to the TiDB Cloud console to create the changefeed:

    - ID in Kafka Advertised Listener Pattern
    - Service Attachment
    - Bootstrap Ports
```

### Network

Ensure that your TiDB cluster can connect to the Apache Kafka service. You can choose one of the following connection methods:

> - Private Connect: ideal for avoiding VPC CIDR conflicts and meeting security compliance, but incurs additional [Private Data Link Cost](/tidbcloud/tidb-cloud-billing-ticdc-rcu.md#private-data-link-cost).
> - VPC Peering: suitable as a cost-effective solution, but requires managing potential VPC CIDR conflicts and security considerations.
> - Public IP: suitable for a quick setup.

Reviewer: ```suggestion
- Private Connect: ideal for avoiding VPC CIDR conflicts and meeting security compliance, but incurs additional [Private Data Link Cost](/tidbcloud/tidb-cloud-billing-ticdc-rcu.md#private-data-link-cost).
- VPC Peering: a cost-effective option, but requires managing potential VPC CIDR conflicts and security considerations.
- Public IP: suitable for a quick setup.
```

#### Private Connect

> Private Connect leverages **Private Link** or **Private Service Connect** technologies provided by cloud vendors. It allows the resources in your VPC to connect to services in other VPCs using private IP addresses, as if those services were hosted directly in your VPC. 
> 
> Currently, TiDB Cloud only supports Private Connect to generic Kafka. No special integration with MSK and Confluent Kafka, or any others.

Reviewer: ```suggestion
Private Connect leverages **Private Link** or **Private Service Connect** technologies from cloud providers to enable resources in your VPC to connect to services in other VPCs using private IP addresses, as if those services were hosted directly within your VPC.

Currently, TiDB Cloud supports Private Connect for generic Kafka only. It does not include special integration with MSK, Confluent Kafka, or other services.
```

> - If your Apache Kafka service is set up in AWS, follow [Set Up Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) to make sure the network connection is set up properly. After you set it up, you can get the following information back to the TiDB Cloud console to create the changefeed:
> 
>     - ID in Kafka Advertised Listener Pattern
>     - Endpoint Service Name
>     - Bootstrap Ports
> 
> - If your Apache Kafka service is set up in Google Cloud, follow [Set Up Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) to make sure the network connection is set up properly. After you set it up, you can get the following information back to the TiDB Cloud console to create the changefeed:
> 
>     - ID in Kafka Advertised Listener Pattern
>     - Service Attachment
>     - Bootstrap Ports

Reviewer: ```suggestion
- If your Apache Kafka service is hosted on AWS, follow [Set Up Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) to ensure that the network connection is properly configured. After setup, provide the following information in the TiDB Cloud console to create the changefeed:

    - ID in Kafka Advertised Listener Pattern
    - Endpoint Service Name
    - Bootstrap Ports

- If your Apache Kafka service is hosted on Google Cloud, follow [Set Up Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) to ensure that the network connection is properly configured. After setup, provide the following information in the TiDB Cloud console to create the changefeedd:

    - ID in Kafka Advertised Listener Pattern
    - Service Attachment
    - Bootstrap Ports
```

### Network

Ensure that your TiDB cluster can connect to the Apache Kafka service. You can choose one of the following connection methods:

- Private Connect: ideal for avoiding VPC CIDR conflicts and meeting security compliance, but incurs additional [Private Data Link Cost](/tidbcloud/tidb-cloud-billing-ticdc-rcu.md#private-data-link-cost).
- VPC Peering: suitable as a cost-effective option, but requires managing potential VPC CIDR conflicts and security considerations.
- Public IP: suitable for a quick setup.

<SimpleTab>
<div label="Private Connect">

Private Connect leverages **Private Link** or **Private Service Connect** technologies from cloud providers to enable resources in your VPC to connect to services in other VPCs using private IP addresses, as if those services were hosted directly within your VPC.

Currently, TiDB Cloud supports Private Connect for generic Kafka only. It does not include special integration with MSK, Confluent Kafka, or other services.

> - If your Apache Kafka service is hosted on AWS, follow [Set Up Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) to ensure that the network connection is properly configured. After setup, provide the following information in the TiDB Cloud console to create the changefeed:

Reviewer: ```suggestion
- If your Apache Kafka service is hosted in AWS, follow [Set Up Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) to ensure that the network connection is properly configured. After setup, provide the following information in the TiDB Cloud console to create the changefeed:
```

    - The ID in Kafka Advertised Listener Pattern
    - The Endpoint Service Name
    - The Bootstrap Ports

> - If your Apache Kafka service is hosted on Google Cloud, follow [Set Up Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) to ensure that the network connection is properly configured. After setup, provide the following information in the TiDB Cloud console to create the changefeed:

Reviewer: ```suggestion
- If your Apache Kafka service is hosted in Google Cloud, follow [Set Up Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) to ensure that the network connection is properly configured. After setup, provide the following information in the TiDB Cloud console to create the changefeed:
```

### Network

Ensure that your TiDB cluster can connect to the Apache Kafka service. You can choose one of the following connection methods:

> - Private Connect: ideal for avoiding VPC CIDR conflicts and meeting security compliance, but incurs additional [Private Data Link Cost](/tidbcloud/tidb-cloud-billing-ticdc-rcu.md#private-data-link-cost).

Author: ```suggestion
- Private Connect (Beta): ideal for avoiding VPC CIDR conflicts and meeting security compliance, but incurs additional [Private Data Link Cost](/tidbcloud/tidb-cloud-billing-ticdc-rcu.md#private-data-link-cost).
```

- VPC Peering: suitable as a cost-effective option, but requires managing potential VPC CIDR conflicts and security considerations.
- Public IP: suitable for a quick setup.

<SimpleTab>

> <div label="Private Connect">

Author: ```suggestion
<div label="Private Connect (Beta)">
```

    You must add the CIDR of the region where your TiDB Cloud cluster is located to the ingress firewall rules. The CIDR can be found on the **VPC Peering** page. Doing so allows the traffic to flow from your TiDB cluster to the Kafka brokers.

#### Public IP

> If you want to provide Public IP access to your Apache Kafka service, you need to assign Public IPs to all you Kafka brokers. It is not recommend to use Public IP in production environment for security consideration. 

Reviewer: ```suggestion
If you want to provide Public IP access to your Apache Kafka service, you need to assign Public IPs to all you Kafka brokers. It is **NOT** recommended to use Public IP in the production environment. 
```

    You must add the CIDR of the region where your TiDB Cloud cluster is located to the ingress firewall rules. The CIDR can be found on the **VPC Peering** page. Doing so allows the traffic to flow from your TiDB cluster to the Kafka brokers.

#### Public IP

> If you want to provide Public IP access to your Apache Kafka service, you need to assign Public IPs to all you Kafka brokers. It is **NOT** recommended to use Public IP in the production environment. 

Reviewer: ```suggestion
If you want to provide Public IP access to your Apache Kafka service, assign Public IPs to all your Kafka brokers. However, using Public IP in a production environment is strongly discouraged. 
```

## Step 2. Configure the changefeed target

> 1. Select **Connectivity Method** by your Apache Kafka Service setup.
>    1. If you select **VPC Peering** or **Public IP**, fill in your Kafka brokers endpoints. You can use commas `,` to separate multiple endpoints.
>    2. If you select **Private Link**
>       1. Please authorize AWS Account of TiDB Cloud, make sure it can create endpoint for your endpoint service. You can find AWS Account of TiDB Cloud in the tip of the web page.
>       2. Make sure you select the same **Number of AZs**, **Suggested Kafka Endpoint Service AZs** and fill the same unique ID in **Kafka Advertised Listener Pattern** when you [Setup Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) in **Network** section.
>       3. Double-check the **Kafka Advertised Listener Pattern** by clicking the button **Check usage and generate**, which will show message to help you validate the unique ID.
>       4. Fill the **Endpoint Service Name** which is configured in [Setup Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md)
>       5. Fill the **Boostrap Ports**, suggest at least one port for one AZ. You can use commas `,` to separate multiple ports.
>    3. If you select **Private Service Connect**
>       1. Make sure you fill the same unique ID in **Kafka Advertised Listener Pattern** when you [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) in **Network** section.
>       2. Double-check the **Kafka Advertised Listener Pattern** by clicking the button **Check usage and generate**, which will show message to help you validate the unique ID.
>       3. Fill the **Service Attachment** which is configured in [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md)
>       4. Fill the **Boostrap Ports**, suggest provide more than one ports. You can use commas `,` to separate multiple ports.
> 2. Select an **Authentication** option according to your Kafka authentication configuration.

Reviewer: ```suggestion
1. Select **Connectivity Method** by your Apache Kafka Service setup.

    - If you select **VPC Peering** or **Public IP**, fill in your Kafka brokers endpoints. You can use commas `,` to separate multiple endpoints.

    - If you select **Private Link**, do the following:

        1. Authorize the AWS Account of TiDB Cloud. Make sure it can create the endpoint for your endpoint service. You can find the AWS Account of TiDB Cloud from the tip of the web page.
        2. Make sure you select the same **Number of AZs** and **Suggested Kafka Endpoint Service AZs**, and fill the same unique ID in **Kafka Advertised Listener Pattern** when you [set up self hosted Kafka Private Link service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) in the **Network** section.
        3. Click **Check usage and generate** to double-check the **Kafka Advertised Listener Pattern**, which will show message to help you validate the unique ID.
        4. Fill the **Endpoint Service Name** which is configured in [Setup Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md).
        5. Fill the **Boostrap Ports**. It is recommended that you set at least one port for one AZ. You can use commas `,` to separate multiple ports.

    - If you select **Private Service Connect**, do the following:
    
        1. Make sure you fill the same unique ID in **Kafka Advertised Listener Pattern** when you [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) in **Network** section.
        2. Click **Check usage and generate** to double-check the **Kafka Advertised Listener Pattern**.  A message is shown to help you validate the unique ID.
        3. Fill the **Service Attachment** that you have configured in [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md)
        4. Fill the **Boostrap Ports**. It is recommended that you provide more than one ports. You can use commas `,` to separate multiple ports.
      
2. Select an **Authentication** option according to your Kafka authentication configuration.
```

    - If your Kafka does not require authentication, keep the default option **Disable**.
    - If your Kafka requires authentication, select the corresponding authentication type, and then fill in the **user name** and **password** of your Kafka account for authentication.

3. Select your **Kafka Version**. If you do not know that, use Kafka V2.
4. Select a desired **Compression** type for the data in this changefeed.
5. Enable the **TLS Encryption** option if your Kafka has enabled TLS encryption and you want to use TLS encryption for the Kafka connection.
6. Click **Validate Connection and Next** to test the network connection, if all is well it will go to the next page.

> > **Note:**
> >  If you select **Private Link** or **Private Service Connect** as network connectivity method. There will be extra steps compare to **Public IP** and **VPC Peering**.
> > 1. After you click the button, we will try to create endpoint in TiDB Cloud side for **Private Link** or **Private Service Connect**. It may take several minutes.
> > 2. After endpoint creation done, you need to accept the connection request in cloud vendor console with you account login.
> > 3. Then go back you TiDB Cloud console to confirm you have already accepted the connection request, then it will test connection and navigate to next page if everything OK.

Reviewer: 下文中的 click the button，是指哪个 button？
```suggestion

If you select **Private Link** or **Private Service Connect** as the network connectivity method, you need to perform the following extra steps.

1. After you click the button, TiDB Cloud creates the endpoint for **Private Link** or **Private Service Connect**. It might take several minutes.
2. After the endpoint is created, you need to log in the cloud vendor console and accept the connection request.
3. Go back your TiDB Cloud console to confirm that you have already accepted the connection request. TiDB Cloud will test the connection and navigate to the next page if everything works.
```

## Step 2. Configure the changefeed target

1. Select **Connectivity Method** by your Apache Kafka Service setup.

    - If you select **VPC Peering** or **Public IP**, fill in your Kafka brokers endpoints. You can use commas `,` to separate multiple endpoints.

    - If you select **Private Link**, do the following:

>         1. Authorize the AWS Account of TiDB Cloud. Make sure it can create the endpoint for your endpoint service. You can find the AWS Account of TiDB Cloud from the tip of the web page.

Reviewer: ```suggestion
        1. Authorize the TiDB Cloud AWS account to create an endpoint for your endpoint service. The TiDB Cloud AWS account ID is provided in the tip on the web page.
```

        2. Make sure you select the same **Number of AZs** and **Suggested Kafka Endpoint Service AZs**, and fill the same unique ID in **Kafka Advertised Listener Pattern** when you [set up self hosted Kafka Private Link service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) in the **Network** section.
        3. Fill the **Endpoint Service Name** which is configured in [Setup Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md).
        4. Fill the **Boostrap Ports**. It is recommended that you set at least one port for one AZ. You can use commas `,` to separate multiple ports.

    - If you select **Private Service Connect**, do the following:
    

>         1. Make sure you fill the same unique ID in **Kafka Advertised Listener Pattern** when you [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) in **Network** section.

Reviewer: ```suggestion
        1. Ensure that you fill the same unique ID in **Kafka Advertised Listener Pattern** when you [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) in **Network** section.
```

        2. Fill the **Service Attachment** that you have configured in [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md)
        3. Fill the **Boostrap Ports**. It is recommended that you provide more than one ports. You can use commas `,` to separate multiple ports.
      
2. Select an **Authentication** option according to your Kafka authentication configuration.
    - If your Kafka does not require authentication, keep the default option **Disable**.
    - If your Kafka requires authentication, select the corresponding authentication type, and then fill in the **user name** and **password** of your Kafka account for authentication.

3. Select your **Kafka Version**. If you do not know that, use Kafka V2.

> 4. Select a desired **Compression** type for the data in this changefeed.

Reviewer: ```suggestion
4. Select a **Compression** type for the data in this changefeed.
```

5. Enable the **TLS Encryption** option if your Kafka has enabled TLS encryption and you want to use TLS encryption for the Kafka connection.
6. Click **Validate Connection and Next** to test the network connection. If the test is successful, you will be directed to the next page.

If you select **Private Link** or **Private Service Connect** as the network connectivity method, follow these additional steps:

> 1. After clicking the button, TiDB Cloud creates the endpoint for **Private Link** or **Private Service Connect**, which might take several minutes.

Reviewer: ```suggestion
1. After clicking **Next**, TiDB Cloud creates the endpoint for **Private Link** or **Private Service Connect**, which might take several minutes.
```

> 1. After clicking the button, TiDB Cloud creates the endpoint for **Private Link** or **Private Service Connect**, which might take several minutes.

Reviewer: Click whic button?

Author: "Next"

> 6. Click **Validate Connection and Next** to test the network connection, if all is well it will go to the next page.
> 
> If you select **Private Link** or **Private Service Connect** as the network connectivity method, you need to perform the following extra steps.
> 
> 1. After you click the button, TiDB Cloud creates the endpoint for **Private Link** or **Private Service Connect**. It might take several minutes.
> 2. After the endpoint is created, you need to log in the cloud vendor console and accept the connection request.
> 3. Go back to the [TiDB Cloud console](https://tidbcloud.com) to confirm that you have already accepted the connection request. TiDB Cloud will test the connection and navigate to the next page if everything works.

Reviewer: ```suggestion
6. Click **Validate Connection and Next** to test the network connection. If the test is successful, you will be directed to the next page.

If you select **Private Link** or **Private Service Connect** as the network connectivity method, follow these additional steps:

1. After clicking the button, TiDB Cloud creates the endpoint for **Private Link** or **Private Service Connect**, which might take several minutes.
2. Once the endpoint is created, log in to your cloud provider console and accept the connection request.
3. Return to the [TiDB Cloud console](https://tidbcloud.com) to confirm that you have accepted the connection request. TiDB Cloud will test the connection and proceed to the next page if the test is successful.
```

2. Select an **Authentication** option according to your Kafka authentication configuration.

    - If your Kafka does not require authentication, keep the default option **Disable**.
    - If your Kafka requires authentication, select the corresponding authentication type, and then fill in the **user name** and **password** of your Kafka account for authentication.

3. Select your **Kafka Version**. If you do not know which one to use, use **Kafka v2**.
4. Select a **Compression** type for the data in this changefeed.
5. Enable the **TLS Encryption** option if your Kafka has enabled TLS encryption and you want to use TLS encryption for the Kafka connection.
6. Click **Next** to test the network connection. If the test succeeds, you will be directed to the next page.

</div>
<div label="Private Link">

1. In **Connectivity Method**, select **Private Link**.
2. Authorize the [AWS Principal](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html#principal-accounts) of TiDB Cloud to create an endpoint for your endpoint service. The AWS Principal is provided in the tip on the web page.

> 3. Make sure you select the same **Number of AZs** and **Suggested Kafka Endpoint Service AZs**, and fill the same unique ID in **Kafka Advertised Listener Pattern** when you [set up self hosted Kafka Private Link service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) in the **Network** section.

Author: ```suggestion
3. Make sure you select the same **Number of AZs** and **AZ IDs of Kafka Deployment**, and fill the same unique ID in **Kafka Advertised Listener Pattern** when you [set up self hosted Kafka Private Link service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md) in the **Network** section.
```

4. Fill the **Endpoint Service Name** which is configured in [Setup Self Hosted Kafka Private Link Service in AWS](/tidb-cloud/setup-self-hosted-kafka-pls.md).
5. Fill the **Bootstrap Ports**. It is recommended that you set at least one port for one AZ. You can use commas `,` to separate multiple ports.
6. Select an **Authentication** option according to your Kafka authentication configuration.

    - If your Kafka does not require authentication, keep the default option **Disable**.
    - If your Kafka requires authentication, select the corresponding authentication type, and then fill in the **user name** and **password** of your Kafka account for authentication.

7. Select your **Kafka Version**. If you do not know which one to use, use **Kafka v2**.
8. Select a **Compression** type for the data in this changefeed.
9. Enable the **TLS Encryption** option if your Kafka has enabled TLS encryption and you want to use TLS encryption for the Kafka connection.
10. Click **Next** to test the network connection. If the test succeeds, you will be directed to the next page.

> 11. TiDB Cloud creates the endpoint for **Private Link** or **Private Service Connect**, which might take several minutes.

Author: ```suggestion
11. TiDB Cloud creates the endpoint for **Private Link**, which might take several minutes.
```

3. Fill the **Service Attachment** that you have configured in [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md)
4. Fill the **Bootstrap Ports**. It is recommended that you provide more than one port. You can use commas `,` to separate multiple ports.      
5. Select an **Authentication** option according to your Kafka authentication configuration.

    - If your Kafka does not require authentication, keep the default option **Disable**.
    - If your Kafka requires authentication, select the corresponding authentication type, and then fill in the **user name** and **password** of your Kafka account for authentication.

6. Select your **Kafka Version**. If you do not know which one to use, use **Kafka v2**.
7. Select a **Compression** type for the data in this changefeed.
8. Enable the **TLS Encryption** option if your Kafka has enabled TLS encryption and you want to use TLS encryption for the Kafka connection.
9. Click **Next** to test the network connection. If the test succeeds, you will be directed to the next page.

> 10. TiDB Cloud creates the endpoint for **Private Link** or **Private Service Connect**, which might take several minutes.

Author: ```suggestion
10. TiDB Cloud creates the endpoint for **Private Service Connect**, which might take several minutes.
```

9. Enable the **TLS Encryption** option if your Kafka has enabled TLS encryption and you want to use TLS encryption for the Kafka connection.
10. Click **Next** to test the network connection. If the test succeeds, you will be directed to the next page.
11. TiDB Cloud creates the endpoint for **Private Link**, which might take several minutes.
12. Once the endpoint is created, log in to your cloud provider console and accept the connection request.
13. Return to the [TiDB Cloud console](https://tidbcloud.com) to confirm that you have accepted the connection request. TiDB Cloud will test the connection and proceed to the next page if the test succeeds.

</div>
<div label="Private Service Connect">

1. In **Connectivity Method**, select **Private Service Connect**.    
2. Ensure that you fill the same unique ID in **Kafka Advertised Listener Pattern** when you [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md) in **Network** section.
3. Fill the **Service Attachment** that you have configured in [Setup Self Hosted Kafka Private Service Connect in Google Cloud](/tidb-cloud/setup-self-hosted-kafka-psc.md)

> 4. Fill the **Bootstrap Ports**. It is recommended that you provide more than one port. You can use commas `,` to separate multiple ports.      

Reviewer 2: ```suggestion
4. Fill in the **Bootstrap Ports**. It is recommended that you provide more than one port. You can use commas `,` to separate multiple ports.      
```

        For changelogs of non-row events, such as Create Schema Event, you can specify a topic name in the **Default Topic Name** field. The changefeed will create a topic accordingly to collect such changelogs.

    - **Distribute changelogs by database to Kafka Topics**

        If you want the changefeed to create a dedicated Kafka topic for each database, choose this mode. Then, all Kafka messages of a database are sent to a dedicated Kafka topic. You can customize topic names of databases by setting a topic prefix and a suffix.

        For changelogs of non-row events, such as Resolved Ts Event, you can specify a topic name in the **Default Topic Name** field. The changefeed will create a topic accordingly to collect such changelogs.

    - **Send all changelogs to one specified Kafka Topic**

        If you want the changefeed to create one Kafka topic for all changelogs, choose this mode. Then, all Kafka messages in the changefeed will be sent to one Kafka topic. You can define the topic name in the **Topic Name** field.

> 7. In the **Partition Distribution** area, you can decide which partition a Kafka message will be sent to. You can define **a single partition dispatcher for all tables**, or you can define **different partition dispatcher for different tables**. We provide 4 types of dispatcher:

Reviewer: ```suggestion
7. In the **Partition Distribution** area, you can decide which partition a Kafka message will be sent to. You can define **a single partition dispatcher for all tables**, or **different partition dispatchers for different tables**. TiDB Cloud provides four types of dispatchers:
```

    - **Distribute changelogs by primary key or index value to Kafka partition**

        If you want the changefeed to send Kafka messages of a table to different partitions, choose this distribution method. The primary key or index value of a row changelog will determine which partition the changelog is sent to. This distribution method provides a better partition balance and ensures row-level orderliness.

    - **Distribute changelogs by table to Kafka partition**

        If you want the changefeed to send Kafka messages of a table to one Kafka partition, choose this distribution method. The table name of a row changelog will determine which partition the changelog is sent to. This distribution method ensures table orderliness but might cause unbalanced partitions.

>    - **Distribute changelogs by timestamp to Kafka partition**
> 
>      If you want the changefeed to send Kafka messages to different Kafka partitions randomly, choose this distribution method. The commitTs of a row changelog will determine which partition the changelog is sent to. This distribution method provides a better partition balance and ensures orderliness in each partition. However, multiple changes of a data item might be sent to different partitions and the consumer progress of different consumers might be different, which might cause data inconsistency. Therefore, the consumer needs to sort the data from multiple partitions by commitTs before consuming.
> 
>    - **Distribute changelogs by columns value to Kafka partition**
> 
>      If you want the changefeed to send Kafka messages of a table to different partitions, choose this distribution method. The specified columns values of a row changelog will determine which partition the changelog is sent to. This distribution method ensures orderliness in each partition and guarantees changelog with same column values will send to the same partition.

Reviewer: ```suggestion
    - **Distribute changelogs by timestamp to Kafka partition**

        If you want the changefeed to send Kafka messages to different Kafka partitions randomly, choose this distribution method. The commitTs of a row changelog will determine which partition the changelog is sent to. This distribution method provides a better partition balance and ensures orderliness in each partition. However, multiple changes of a data item might be sent to different partitions and the consumer progress of different consumers might be different, which might cause data inconsistency. Therefore, the consumer needs to sort the data from multiple partitions by commitTs before consuming.

    - **Distribute changelogs by column value to Kafka partition**

        If you want the changefeed to send Kafka messages of a table to different partitions, choose this distribution method. The specified column values of a row changelog will determine which partition the changelog is sent to. This distribution method ensures orderliness in each partition and guarantees that the changelog with the same column values is send to the same partition.
```
